cli/cook/cli.py

```diff
@@ -0,0 +1,112 @@
+subparsers = parser.add_subparsers(dest='action')
+
+actions = {'submit': submit.register(subparsers.add_parser),
+           'show': show.register(subparsers.add_parser),
+           'wait': wait.register(subparsers.add_parser)}
+
+
+def load_target_clusters(config, url=None, cluster=None):
+    """Given the config and (optional) url and cluster flags, returns the list of clusters to target"""
+    if cluster and url:
+        raise Exception('You cannot specify both a cluster name and a cluster url at the same time')
+
+    clusters = None
+    config_clusters = config.get('clusters')
+    if url:
+        if urlparse(url).scheme == '':
+            url = 'http://%s' % url
+        clusters = [{'name': url, 'url': url}]
+    elif config_clusters:
+        if cluster:
+            clusters = [c for c in config_clusters if c.get('name') == cluster]
+        else:
+            clusters = [c for c in config_clusters if 'disabled' not in c or not c['disabled']]
+
+    if not clusters:
+        raise Exception('%s\nYour current configuration is:\n%s' %
+                        (colors.failed('You must specify at least one cluster.'), json.dumps(config, indent=2)))
+
+    return clusters
+
+
+def load_config(config_path=None):
+    """Loads the configuration map to use"""
+    if config_path:
+        if os.path.isfile(config_path):
+            with open(config_path) as json_file:
+                config = json.load(json_file)
+        else:
+            raise Exception('The configuration path specified (%s) is not valid' % config_path)
+    else:
+        config = load_first_json_file(DEFAULT_CONFIG_PATHS) or {}
+    config = deep_merge(DEFAULT_CONFIG, config)
+    logging.debug('using configuration: %s' % config)
+    return config
+
+
+def run(args):
+    """
+    Main entrypoint to the cook scheduler CLI. Loads configuration files, 
+    processes global command line arguments, and calls other command line 
+    sub-commands (actions) if necessary.
+    """
+    args = vars(parser.parse_args(args))
+    util.silent = args.pop('silent')
+    verbose = args.pop('verbose') and not util.silent
+
+    log_format = '%(asctime)s [%(levelname)s] %(message)s'
+    if verbose:
+        logging.basicConfig(format=log_format, level=logging.DEBUG)
+    else:
+        logging.basicConfig(format=log_format, level=logging.FATAL)
+
+    logging.debug('args: %s' % args)
+
+    action = args.pop('action')
+    config_path = args.pop('config')
+    cluster = args.pop('cluster')
+    url = args.pop('url')
+
+    if action is None:
+        parser.print_help()
+    else:
+        config = load_config(config_path)
+        clusters = load_target_clusters(config, url, cluster)
+        http.configure(config)
+        args = {k: v for k, v in args.items() if v is not None}
+        defaults = config.get('defaults')
+        action_defaults = (defaults.get(action) if defaults else None) or {}
+        result = actions[action](clusters, deep_merge(action_defaults, args))
```

Comment:
Is it possible for `action` not be in the `actions` dictionary? If so, should we do some error handling?

Reply:
`argparse` handles this for us:

```bash
$ cs foo
usage: cs [-h] [--cluster CLUSTER] [--url URL] [--config CONFIG] [--silent]
          [--verbose]
          {submit,show,wait} ...
cs: error: argument action: invalid choice: 'foo' (choose from 'submit', 'show', 'wait')
```